kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala

```diff
@@ -148,6 +149,39 @@ trait ProcBuilder {
     proc
   }
 
+  def execCommand(command: String): Int = {
+    val pb = new ProcessBuilder("/bin/sh", "-c", command)
+    var exitValue: Option[Int] = None
+
+    try {
+      val process = pb.start()
+      exitValue = Some(process.waitFor())
+    } catch {
+      case e: Exception =>
+        throw KyuubiSQLException(s"Failed to kill yarn job. command : $command", e.getCause)
+    }
+    exitValue.get
+  }
+
+  def getCommand(appId: String): String = {
+    val hadoopHomeOpt = env.get("HADOOP_HOME").orElse {
+      throw KyuubiSQLException(s"Failed to kill yarn job. HADOOP_HOME is not set! " +
+        "For more detail information on installing and configuring Yarn, please visit " +
+        "https://kyuubi.apache.org/docs/stable/deployment/on_yarn.html")
+    }.map { hadoopHome =>
+      Paths.get(hadoopHome, "bin", "yarn").toAbsolutePath.toFile.getCanonicalPath
+    }.get
+    hadoopHomeOpt + " application -kill " + appId
+  }
+
+  val YARN_APP_NAME_REGEX: Regex = "application_\\d+_\\d+".r
+
+  def killApplication(line: String = lastRowOfLog): Unit =
+    YARN_APP_NAME_REGEX.findFirstIn(line) match {
+      case Some(appId) if line.contains("state: ACCEPTED") => execCommand(getCommand(appId))
```

Comment:
It is possible, although rare, that the last row of log reports state to be NEW/NEW_SAVING/SUBMITTED when engine is submitted in YARN Cluster mode and `spark.yarn.submit.waitAppCompletion` is `false`.